dev_scripts/qa.py

```diff
@@ -0,0 +1,885 @@
+        """
+        # Remove '#' from header
+        anchor = re.sub("^[#]+", "", self.heading_title)
+        # Format
+        anchor = anchor.strip().lower()
+        # Convert spaces to dashes
+        anchor = anchor.replace(" ", "-")
+        # Remove non-alphanumeric (except dash and underscore)
+        anchor = re.sub("[^a-zA-Z\-_]", "", anchor)
+
+        return anchor
+
+    def diff(self, source):
+        """Return a diff between the section in the docs and the stored one."""
+        source = source.splitlines(keepends=True)
+        content = self.content.splitlines(keepends=True)
+        diff = difflib.unified_diff(
+            source, content, fromfile=self.md_path, tofile="qa.py"
+        )
+        return diff
+
+
+class QABase(abc.ABC):
+    """Base class for the QA tasks."""
+
+    platforms = {}
+
+    REF_QA = Reference("RELEASE.md", content=CONTENT_QA)
+    REF_QA_SCENARIOS = Reference("RELEASE.md", content=CONTENT_QA_SCENARIOS)
+
+    # The following class method is available since Python 3.6. For more details, see:
+    # https://docs.python.org/3.6/whatsnew/3.6.html#pep-487-simpler-customization-of-class-creation
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+        # NOTE: Not all subclasses correspond to valid platforms. Some subclasses may
+        # provide some basic functionality for an OS type or distro, but not a specific
+        # version.
+        if cls.get_id():
+            cls.platforms[cls.get_id()] = cls
+
+    def __init__(self, try_auto=False, skip_manual=False, debug=False):
+        self.debug = debug
+        self.try_auto = try_auto
+        self.skip_manual = skip_manual
+        self.src = (
+            subprocess.run(
+                [
+                    "git",
+                    "rev-parse",
+                    "--show-toplevel",
+                ],
+                stdout=subprocess.PIPE,
+            )
+            .stdout.decode()
+            .strip("\n")
+        )
+
+    def task(*msgs, ref=None, auto=False):
+        """Decorator for running a task automatically.
+
+        This decorator does the following:
+        * Check if the user has asked to run the tasks automatically.
+          - If not, ask the user if they want to try it out.
+          - If yes, run the task (decorated function).
+        * If an exception occurs, allow the user to skip the task, or stop the
+          execution.
+
+        Args:
+            msgs: A list of messages that we should show to the user, one message per
+              line.
+            ref: A link that can be used as a reference for a specific task.
+        """
+
+        def decorator(func):
+            def inner(self, *args, **kwargs):
+                # If the reference is actually defined in subclasses, grab it from
+                # `self`.
+                if isinstance(ref, str):
+                    _ref = getattr(self, ref)
+                else:
+                    _ref = ref
```

Comment:
Sigh, I'm not proud of this code. An example where the problem manifests is here:

https://github.com/freedomofpress/dangerzone/blob/f0c3c82f25e77b3e59a30a776e926a5a840fbcc0/dev_scripts/qa.py#L699-L700

If you look carefully, I use `ref="REF_BUILD"`, instead of `ref=REF_BUILD`. The reason is that during class creation, `REF_BUILD = None`. To accomodate for the `REF_BUILD` that the subclasses later define, I had to resort to a dynamic approach :disappointed: .